src/operator/tensor/elemwise_binary_op.h

```diff
@@ -25,363 +25,798 @@
-  Stream<xpu> *s = ctx.get_stream<xpu>();
-  int size = static_cast<int>((outputs[0].Size() + DataType<DType>::kLanes - 1)
-    /DataType<DType>::kLanes);
-  DType* lgrad_dptr = outputs[0].dptr<DType>();
-  DType* rgrad_dptr = outputs[1].dptr<DType>();
-  DType* ograd_dptr = inputs[0].dptr<DType>();
-  if (std::is_same<LOP, mshadow_op::identity>::value && req[0] == kWriteInplace) {
-    CHECK_EQ(ograd_dptr, lgrad_dptr);
-  } else if (req[0] != kNullOp) {
-    MXNET_ASSIGN_REQ_SWITCH(req[0], Req,
-      {Kernel<BinaryOpBackwardUseNone<LOP, Req>, xpu>::Launch(s, size, lgrad_dptr,
-        ograd_dptr);});
+/*! Gather binary operator functions into BinaryOp class */
+class ElemwiseBinaryOp : public OpBase {
+ public:
+  template<typename OP, int Req>
+  struct BinaryOpBackwardUseNone {
+    template<typename DType>
+    MSHADOW_XINLINE static void Map(int i, DType *igrad, const DType *ograd) {
+      KERNEL_ASSIGN(igrad[i], Req, OP::Map(ograd[i]));
+    }
+    template<typename DType>
+    MSHADOW_XINLINE static DType Map(const DType ograd) {
+      return OP::Map(ograd);
+    }
+  };
+
+  template<typename OP, int Req>
+  struct BinaryOpBackwardUseIn {
+    template<typename DType>
+    MSHADOW_XINLINE static void Map(int i, DType *igrad,
+                                    const DType *ograd, const DType *lhs, const DType *rhs) {
+      KERNEL_ASSIGN(igrad[i], Req, ograd[i] * OP::Map(lhs[i], rhs[i]));
+    }
+  };
+
+  /*! \brief For sparse, assume missing rvalue is 0 */
+  template<typename OP, int Req>
+  struct BinaryOpMissingRValue {
+    template<typename DType>
+    MSHADOW_XINLINE static void Map(int i, DType *out, const DType *lhs) {
+      KERNEL_ASSIGN(out[i], Req, OP::Map(lhs[i], DType(0)));
+    }
+  };
+
+  /*! \brief For sparse, assume missing lvalue is 0 */
+  template<typename OP, int Req>
+  struct BinaryOpMissingLValue {
+    template<typename DType>
+    MSHADOW_XINLINE static void Map(int i, DType *out, const DType *rhs) {
+      KERNEL_ASSIGN(out[i], Req, OP::Map(DType(0), rhs[i]));
+    }
+  };
+
+ private:
+  /*! \brief Fill dense output block with a single scalar value */
+  template<typename xpu, typename DType>
+  static inline void Fill(mshadow::Stream<xpu> *s,
+                          const DType val,
+                          const OpReqType req,
+                          const TBlob& blob) {
+    using namespace mxnet_op;
+    using namespace mshadow::expr;
+    MXNET_ASSIGN_REQ_SWITCH(req, Req, {
+      Kernel<MapSetToScalar<Req>, xpu>::Launch(s, blob.Size(), blob.dptr<DType>(), val);
+    });
   }
-  if (std::is_same<ROP, mshadow_op::identity>::value && req[1] == kWriteInplace) {
-    CHECK_EQ(ograd_dptr, rgrad_dptr);
-  } else if (req[1] != kNullOp) {
-    MXNET_ASSIGN_REQ_SWITCH(req[1], Req,
-      {Kernel<BinaryOpBackwardUseNone<ROP, Req>, xpu>::Launch(s, size, rgrad_dptr,
-        ograd_dptr);});
+
+  /*! \brief Fill contiguous dense output rows with value computed from 0 lhs and 0 rhs input */
```

Comment:
For some dense results, you can just fill with a single value which is calculated donce